File: problem2/SymmetricTree.java
// Time Complexity : O(n)
// Space Complexity : O(h)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No
package problem2;

public class SymmetricTree {
	public boolean isSymmetric(TreeNode root) {
		if (root == null) {
			return true;
		}

		return helper(root.left, root.right);
	}

	public boolean helper(TreeNode node1, TreeNode node2) {
		if (node1 == null && node2 == null) {
			return true;
		}

		if (node1 == null || node2 == null) {
			return false;
		}

		return node1.val == node2.val && helper(node1.left, node2.right) && helper(node1.right, node2.left);
	}

	public static void main(String[] args) {
		SymmetricTree obj = new SymmetricTree();

		TreeNode root = new TreeNode(1);
		root.left = new TreeNode(2);
		root.right = new TreeNode(2);
		root.left.left = new TreeNode(3);
		root.left.right = new TreeNode(4);
		root.right.left = new TreeNode(4);
		root.right.right = new TreeNode(3);

		System.out.println("Is the tree symmetric? " + (obj.isSymmetric(root) == true ? "Yes" : "No"));
	}

}

File: SymmetricTree.cpp
// Approach 1 - void based DFS
// Time Complexity - O(n), since we are processing a total of 'n' nodes in the input tree.
// Space Complexity - O(h) - recursive stack
// Problems Faced - No!
// It runs on leetcode.
class Solution {
    bool flag;
    private:
    void dfs(TreeNode* nodeL, TreeNode* nodeR){
        if(nodeL == NULL && nodeR == NULL)
            return;
        if(nodeL == NULL || nodeR == NULL){
            flag = false;
            return;
        }
        if(nodeL->val != nodeR->val){
            flag = false;
            return;
        }
        
        dfs(nodeL->left, nodeR->right);
        if(flag)
            dfs(nodeL->right, nodeR->left);
    }
public:
    bool isSymmetric(TreeNode* root) {
        flag = true;
        dfs(root->left, root->right);
        return flag;
    }
};

// Approach 2 - BFS
// Time Complexity - O(n), since we are processing a total of 'n' nodes in the input tree.
// Space Complexity - O(n) - Queue space
// Problems Faced - No!
// It runs on leetcode.
class Solution {
public:
    bool isSymmetric(TreeNode* root) {
        queue<TreeNode*> q;
        q.push(root->left);
        q.push(root->right);
        
        while(!q.empty()){
            TreeNode* left = q.front(); q.pop();
            TreeNode* right = q.front(); q.pop();
            
            if(left == NULL && right == NULL) continue;
            if(left == NULL || right == NULL || left->val != right->val)
                return false;
            q.push(left->left); q.push(right->right);
            q.push(left->right); q.push(right->left);
        }
        return true;
    }
};